Trans with master eclstate

blattms opened the pull request:
This incorporates the lastest changes to master from my PR OPM/opm-simulators#2243. It also needs
the changes to opm-grid. A backport of OPM/opm-grid#433 to the branch off time at Dec 10, 2019 can
be found in branch[noecl_base](https://github.com/blattms/opm-grid/tree/noecl_base).

Tested on SPE9. Not sure yet whether Norne works (there might be an issue if there is a GeoModifier
event as we might need to update more data).

blattms committed:
Pass a pointer to the EclipseGrid to CpGrid::processEclipseGrid.

... instead of a reference This is needed as we only want to read the
full deck and construct the EclipseGrid only on the master process
with rank 0. Hence all other ranks will pass a nullptr to this
function. This will be possible with this move from a reference to a
pointer.

blattms committed:
Make ECLTransmissibility work EclState only on master.

To accomplish this we need to change the caclulation of the
transmissibility multipliers. In a parallel these are now
computed globally for the loadbalancing and then broadcasted and
cached in the vanguard. Later these are moved to the
EclTransmissibilty class and used in the next
EclTransmissibilty::update call. Later these will be recomputed.
(At that stage there might still be an update of MultTrans missing.)

akva2 commented on ebos/eclcpgridvanguard.hh:
```diff
@@ -276,9 +278,30 @@ protected:
         // equilGrid_being a shallow copy only the global view.
         if (mpiRank == 0)
         {
+            const auto& gridProps = this->eclState(false).get3DProperties(); // this is hit
```
please remove the hit comment and let it assert.

blattms replied:
done

akva2 commented on ebos/eclcpgridvanguard.hh:
```diff
@@ -276,9 +278,30 @@ protected:
+                                    /*flipNormals=*/false,
+                                    /*clipZ=*/false,
+                                    porv,
+                                    this->eclState(false).getInputNNC()); // this is hit
```
ditto.

blattms replied:
done

akva2 commented on ebos/eclshallowgrid.hh:
```diff
@@ -0,0 +1,133 @@
+#include <vector>
+namespace Opm
+{
+/// \brief Utility class that replicates cartesian grid data for prosity
```
porosity

blattms replied:
done

blattms commented:
Thanks. I amended the last commit.

akva2 commented on ebos/eclcpgridvanguard.hh:
```diff
@@ -276,9 +278,30 @@ protected:
+                                    /*flipNormals=*/false,
+                                    /*clipZ=*/false,
+                                    porv,
+                                    this->eclState(true).getInputNNC()); // this is hit
```
missed it first time. one more..

blattms commented:
Norne is crashing.

blattms commented:
or better: fails in an assertion
```
(gdb) up
#5  0x00005555573efca4 in Opm::EclOutputBlackOilModule<Opm::Properties::TTag::EclFlowProblem>::addRftDataToWells (this=0x555564a5d1d0, wellDatas=..., 
    reportStepNum=7)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/ebos/ecloutputblackoilmodule.hh:871
871                         const size_t index = simulator_.vanguard().eclState(true).getInputGrid().getGlobalIndex(i, j, k);
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff4be5535 in __GI_abort () at abort.c:79
#2  0x00007ffff4be540f in __assert_fail_base (fmt=0x7ffff4d47ee0 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=0x555558dd5c99 "myRank == 0", 
    file=0x555558dcee58 "/home/mblatt/src/dune/opm-2.6/opm-simulators/ebos/eclbasevanguard.hh", line=398, function=<optimized out>) at assert.c:92
#3  0x00007ffff4bf3102 in __GI___assert_fail (assertion=0x555558dd5c99 "myRank == 0", 
    file=0x555558dcee58 "/home/mblatt/src/dune/opm-2.6/opm-simulators/ebos/eclbasevanguard.hh", line=398, 
    function=0x555558de58e0 <Opm::EclBaseVanguard<Opm::Properties::TTag::EclFlowProblem>::eclState(bool) const::__PRETTY_FUNCTION__> "const Opm::EclipseState& Opm::EclBaseVanguard<TypeTag>::eclState(bool) const [with TypeTag = Opm::Properties::TTag::EclFlowProblem]") at assert.c:101
#4  0x000055555739cf86 in Opm::EclBaseVanguard<Opm::Properties::TTag::EclFlowProblem>::eclState (this=0x555559edbc50, assert=true)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/ebos/eclbasevanguard.hh:398
#5  0x00005555573efca4 in Opm::EclOutputBlackOilModule<Opm::Properties::TTag::EclFlowProblem>::addRftDataToWells (this=0x555564a5d1d0, wellDatas=..., 
    reportStepNum=7) at /home/mblatt/src/dune/opm-2.6/opm-simulators/ebos/ecloutputblackoilmodule.hh:871
#6  0x00005555573b04cc in Opm::EclWriter<Opm::Properties::TTag::EclFlowProblem>::writeOutput (this=0x555564a5cff0, isSubStep=false)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/ebos/eclwriter.hh:362
#7  0x0000555557374d30 in Opm::EclProblem<Opm::Properties::TTag::EclFlowProblem>::writeOutput (this=0x555564a5bac0, verbose=true)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/ebos/eclproblem.hh:1094
#8  0x000055555735cf84 in Opm::SimulatorFullyImplicitBlackoilEbos<Opm::Properties::TTag::EclFlowProblem>::run (this=0x555559cf1a60, timer=...)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/opm/simulators/flow/SimulatorFullyImplicitBlackoilEbos.hpp:249
#9  0x0000555557344331 in Opm::FlowMainEbos<Opm::Properties::TTag::EclFlowProblem>::runSimulator (this=0x7fffffffa9f0, output_cout=false)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/opm/simulators/flow/FlowMainEbos.hpp:487
#10 0x000055555733628d in Opm::FlowMainEbos<Opm::Properties::TTag::EclFlowProblem>::execute (this=0x7fffffffa9f0, argc=2, argv=0x7fffffffd6e8, output_cout=false, 
    output_to_files=true) at /home/mblatt/src/dune/opm-2.6/opm-simulators/opm/simulators/flow/FlowMainEbos.hpp:261
#11 0x00005555572dcbf2 in Opm::flowEbosBlackoilMain (argc=2, argv=0x7fffffffd6e8, outputCout=false, outputFiles=true)
    at /home/mblatt/src/dune/opm-2.6/opm-simulators/flow/flow_ebos_blackoil.cpp:63
#12 0x00005555571ca1bd in main (argc=2, argv=0x7fffffffd6e8) at /home/mblatt/src/dune/opm-2.6/opm-simulators/flow/flow.cpp:463

```
on non-zero ranks during output.

Known issue?

akva2 commented:
no. probably due to updates in mainline that i have not addressed yet.

blattms committed:
Removes usage of EclipseState for all processors during well output.

addRftDataToWells still accessed EclipseState on all processors
when a well git shut.

blattms commented:
I must have introduced a subtle bug that breaks convergence for norne.
